### ISSUE TITLE
在使用做LSTM时有如下报错LstmLayer.cpp  Check failed: input.sequenceStartPositions

### ISSUE THREAD

**Issue by MEYOUTH** [PaddlePaddle/Paddle#8111]
Thank you for contributing to PaddlePaddle. Submitting an issue is a great help for us.
Both Chinese and English issues are welcome.

It's hard to solve a problem when important details are missing.
Before submitting the issue, look over the following criteria before handing your request in.

- [ ] Was there a similar issue submitted or resolved before ? You could search issue in the github.
- [ ] Did you retrieve your issue from widespread search engines ?
- [ ] Is my description of the issue clear enough to reproduce this problem?
   * If some errors occurred, we need details about `how do you run your code?`, `what system do you use?`, `Are you using GPU or not?`, etc.
   * If you use an recording [asciinema](https://asciinema.org/) to show what you are doing to make it happen, that's awesome! We could help you solve the problem more quickly.
- [ ] Is my description of the issue use the github markdown correctly?
   * Please use the proper markdown syntaxes for styling all forms of writing, e.g, source code, error information, etc.
   * Check out [this page](https://guides.github.com/features/mastering-markdown/) to find out much more about markdown.

**Comment by peterzhang2029** [comment]
把你的配置和报错的完整信息发一下吧

**Comment by MEYOUTH** [comment]
I0204 00:43:59.307287  3734 Util.cpp:166] commandline:  --use_gpu=False --trainer_count=1 
I0204 00:44:01.862337  3734 GradientMachine.cpp:94] Initing parameters..
I0204 00:44:12.529839  3734 GradientMachine.cpp:101] Init parameters done.
F0204 00:44:12.904106  3734 LstmLayer.cpp:155] Check failed: input.sequenceStartPositions 
*** Check failure stack trace: ***
    @     0x7f1ca36d212d  google::LogMessage::Fail()
    @     0x7f1ca36d5bdc  google::LogMessage::SendToLog()
    @     0x7f1ca36d1c53  google::LogMessage::Flush()
    @     0x7f1ca36d70ee  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f1ca33ec7e2  paddle::LstmLayer::forward()
    @     0x7f1ca3482ebd  paddle::NeuralNetwork::forward()
    @     0x7f1ca3483bd3  paddle::GradientMachine::forwardBackward()
    @     0x7f1ca36ae004  GradientMachine::forwardBackward()
    @     0x7f1ca32fa9c9  _wrap_GradientMachine_forwardBackward
    @           0x4cb240  PyEval_EvalFrameEx
    @           0x4cfedc  PyEval_EvalCodeEx
    @           0x4c9419  PyEval_EvalFrameEx
    @           0x4cfedc  PyEval_EvalCodeEx
    @           0x4c8314  PyEval_EvalFrameEx
    @           0x4cfedc  PyEval_EvalCodeEx

**Comment by MEYOUTH** [comment]
fc_para_attr = paddle.attr.Param(learning_rate=1e-3)
    lstm_para_attr = paddle.attr.Param(initial_std=0., learning_rate=1.)
    para_attr = [fc_para_attr, lstm_para_attr]
    bias_attr = paddle.attr.Param(initial_std=0., l2_rate=0.)
    relu = paddle.activation.Relu()
    linear = paddle.activation.Linear()
我输入的是一个时序的矩阵
x = paddle.layer.data("x",  type=paddle.data_type.dense_vector(600),height=12,width=50)
    fc1 = paddle.layer.fc(input=x,
                          size=hid_dim*4,
                          act=linear,
                          bias_attr=bias_attr)
    lstm1 = paddle.layer.lstmemory(
        input=fc1, act=relu, bias_attr=bias_attr)
    predict = paddle.layer.fc(input=[fc1, lstm1],size=2,act=paddle.activation.Softmax(),bias_attr=bias_attr,param_attr=para_attr)

**Comment by peterzhang2029** [comment]
Lstm要求你的输入是序列数据 数据类型要使用integer_value_sequence或者dense_vector_sequence来表示序列数据。 详细区别可参照数据类型
 http://www.paddlepaddle.org/docs/develop/documentation/zh/getstarted/concepts/use_concepts_cn.html

**Comment by MEYOUTH** [comment]
是要变成这样吗x = paddle.layer.data("x", type=paddle.data_type.dense_vector_sequence(12),width=50)

**Comment by MEYOUTH** [comment]
官网上做语义分析是data = paddle.layer.data("word",
                             paddle.data_type.integer_value_sequence(input_dim))
    emb = paddle.layer.embedding(input=data, size=emb_dim)
可是我不是做语义的

**Comment by peterzhang2029** [comment]
那你的数据是稠密向量的序列么？

**Comment by MEYOUTH** [comment]
是稠密的序列

**Comment by peterzhang2029** [comment]
按照这样写试试 https://github.com/PaddlePaddle/models/blob/3b75c23b81cc4b72661cfd3a71a1cb5ab8d4cf4a/ltr/lambda_rank.py#L18  input_dim设置为你向量的大小

**Comment by MEYOUTH** [comment]
不是很明白，我的数据是稠密向量的序列，12*50的，12是表征时间，input_dim设为50还是12，那我每个样本12*50是否要转化为1*600吗

**Comment by peterzhang2029** [comment]
1. input_dim设为50
2. 不需要， 每个样本是处理为 [[f, ...], [f, ...], ...] ，[f, ...]表示你的50维向量。https://github.com/PaddlePaddle/models/blob/3b75c23b81cc4b72661cfd3a71a1cb5ab8d4cf4a/ltr/lambda_rank.py#L18  上的注释有说明

**Comment by MEYOUTH** [comment]
我只改了这一句，改成这样data = paddle.layer.data("data",
paddle.data_type.dense_vector_sequence(50)
其他lstm还是一样，还是不行
I0204 04:21:06.376062 60603 GradientMachine.cpp:101] Init parameters done.
F0204 04:21:06.959400 60603 Matrix.cpp:3680] Check failed: label.getSize() == numSamples (30 vs. 420)
label是不是也要改呢

**Comment by MEYOUTH** [comment]
还是不明白，链接里面他的写法好像并不适合做LSTM

**Comment by peterzhang2029** [comment]
你这里的报错应该和LSTM输入没有关系 应该是你的网络配置有问题，你最终的softmax的输入应该是一个向量吧，而LSTM的输出是一个序列，一般是取序列最后的隐变量来作为softmax的输入进行分类。

**Comment by MEYOUTH** [comment]
fc_last = paddle.layer.pooling(input=fc1, pooling_type=paddle.pooling.Max())
    lstm_last = paddle.layer.pooling(input=lstm1, pooling_type=paddle.pooling.Max())
我最后在predict'前加了这个就可以了

**Comment by peterzhang2029** [comment]
pooling也是将序列转为向量， 都是可以的

**Comment by MEYOUTH** [comment]
那我这样写对分类有影响吗

**Comment by peterzhang2029** [comment]
一般的做法是用paddle.layer.last_seq(input=lstm_output) https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/layers.py#L1721 来取得lstm输出的最后一个时间步的隐变量，然后进一步处理，输入到softmax层。单纯的比较两种做法，很难说哪种效果更好，还得根据数据看具体效果。

**Comment by MEYOUTH** [comment]
好的，谢谢您的解答，受益匪浅。我还有个新问题#8112
